### DISCUSSION TITLE
A small tip that might help

### DISCUSSION THREAD

**Discussion by dayunbao** [hako-mikan/sd-webui-regional-prompter#206] [General]
I'm going to be integrating Regional Prompter into my comic book creating app, and was a little confused at first by the Divide Ratio.  I discovered that it seems to ignore parentheses and white space, which is nice since it helps to visually group the different parts of the Divide Ratio. 

![regional_prompter_tip](https://github.com/hako-mikan/sd-webui-regional-prompter/assets/11196050/65067b51-7e06-4445-8eb1-bfefb5ab030b)

Using some spaces and parentheses for grouping the Divide Ratio, along with the visualize option, this `1 (;2, 3, 1) (;1, 2, 1, 1)` makes it much easier to understand what's going on (for me, at least).

So we have :
Row 1 = `1`
Row 2 = `(;2, 3, 1) `
Row 3 = `(;1, 2, 1, 1)`

This way it's also easy to see how the columns are divided in each row.  So Row 2 has 2 columns, where the 1st column is 3 times the width of the 2nd.

Initially, I was also confused about how to specify multiple columns within a row.  My first attempt was Row 2 = `(;2, 1)`, which didn't change anything.  When I added another `, 1`, making the Divide Ratio `(;2, 1, 1) ` then I got 2 columns in the row.  So it seems that there is always 1 implicit column (or row), instead of starting with 0 columns or rows.

I don't know if this helps anyone else, but it helped me a lot.

**Comment by Speusippus** [comment]
I cannot wrap my head around this whole system (not your notational aid, the system itself, with or without your addition). 

Here's a puzzle which I suspect, if I can grasp what's going on here, I may finally get what's going on.

Columns mode, DR 1,1,1 <---this gives me three equally sized columns
Columns mode, DR 1;1;1 <---this gives me three equally sized rows
Rows mode, DR 1,1,1 <---This gives me three equally sized rows
Rows mode, DR 1;1;1 <---This gives me TWO(?!) equally sized ROWS(?!)

That fourth one just completely throws me. Can you or can anyone here help me understand what's going on here? I have hopes that understanding this particular case will help me understand the even more confusing stuff that happens when I try to do more complicated grids. I have read the explanation, and it even seems to make sense when I read it slowly, but then when I try to put into practice what I think I grasped--it doesn't work. I'm very clearly misunderstanding something basic about this.

**Reply by Symbiomatrix** [reply]
In general, commas separate columns and semicolons separate rows. What mode controls is which comes first.
However, for backwards compatibility and simplicity, I've made it so that if only commas are used, it will switch to 1D mode, ie one will create rows, the other will create columns (we used to call it horizontal and vertical split, the new naming scheme seems alright).
So, with that in mind (I've used parenthesis in the explanation loosely for visual grouping, don't read too much into them):
Case 2 is shorthand for (1,1) ; (1,1) ; (1,1) , first splitting to three 33% rows, then each comprising a single 100% cell. Because we're in columns mode, each semicolon creates a new "row block" and that's why we get 3 of them. We were only given a single value for each row, so that value is propagated for the inner columns (doesn't matter here since they're alone but it matters when the values are matched with multiple ADDX regions which haven't been assigned a proportion value).
Case 4 is read as 1 ; (1;1) , meaning two 50% rows within a single column that's 100% of the columns.
Because we're in rows mode, semicolons are internal row separators in a single column. We don't (need to) propagate values, the first internal separator separates the column's relative size to other columns from the inner rows' relative sizes.

**Comment by Speusippus** [comment]
By the way, a note about your post--it looks like parentheses do affect the visualization somehow. 

I put it in "Rows" mode then entered this divide ratio:

1;1;2;2

This creates a row of size 1, a row of size 2, and a row of size 2.

Then I changed it to

(1;1;2;2)

And this instead creates a row of size 1, a row of size 2, and a row of size 1.

**Reply by Symbiomatrix** [reply]
Yeah, that sounds to me more likely unless the code's seen some changes.
To avoid pestering error messages, each proportion value is silently changed to a default if it cannot be converted to a number. Whitespace is okay, fractions are okay, any other text is not.
In the author's original sample the corrupt values are already 1, eg "1) (" so the defaultisation does nothing, but in yours that last "2)" is changed to 1.

**Reply by dayunbao** [reply]
I built my own GUI for Regional Prompter, and use pixel values.  I posted screenshots that show the divide ratios for both modes here https://github.com/hako-mikan/sd-webui-regional-prompter/discussions/218#discussioncomment-6896327.  Maybe that will help.
